### PR TITLE
[Pygen] Add Complete flow for RV32I instructions

### PR DESCRIPTION
**Pull request by aneels3**
Generated assembly file 
image
## Side Note

- Immediate values are not up to the mark because **extend_imm()** function and constraint of **self.imm** variable are not up yet. [RESOLVED] in  <https://github.com/google/riscv-dv/tree/949f666959bfb8c65bc92db71c731523a8dcf1c0> commit.

image

## Generate a simple test
In test directory
`python3 riscv_instr_base_test.py`

### CODE REVIEW

**Comment by ShraddhaDevaiya** [comment]
@googlebot I fixed it.

**Comment by ShraddhaDevaiya** [comment]
@googlebot I consent.

File: pygen/pygen_src/isa/riscv_instr.py [hunk 1]
```diff
@@ -9,23 +9,25 @@
...
-from pygen_src.riscv_instr_pkg import riscv_instr_group_t
+from pygen_src.riscv_instr_pkg import pkg_ins, riscv_reg_t
 from pygen_src.isa import rv32i_instr  # NOQA
+from pygen_src.target.rv32i import riscv_core_setting as rcs
 import random
 from bitstring import BitArray
 import logging
 from copy import copy
 import sys
 import vsc
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:24]: very small nit: can all of the `from...import...` statements be moved under the `import ...` statements?

File: pygen/pygen_src/isa/riscv_instr.py [hunk 2]
```diff
@@ -46,12 +48,12 @@ def __init__(self):
...
-        self.rs1 = None
-        self.rd = None
-        self.imm = BitArray(hex="0x00000000")
+        self.rs2 = vsc.rand_enum_t(riscv_reg_t)
+        self.rs1 = vsc.rand_enum_t(riscv_reg_t)
+        self.rd = vsc.rand_enum_t(riscv_reg_t)
+        self.imm = vsc.rand_uint32_t()
 
-        self.imm_mask = BitArray(hex="0xffffffff")
+        self.imm_mask = 0xffffffff
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:56]: why is this changed to no longer be a BitArray?

File: pygen/pygen_src/isa/riscv_instr.py [hunk 3]
```diff
@@ -62,20 +64,15 @@ def __init__(self):
...
         self.has_rs2 = 1
         self.has_rd = 1
         self.has_imm = 1
-
-        # Fields Added for debugging These fields are actually from a different files.
-        self.unsupported_instr = []
-        self.XLEN = 32
-        self.supported_isa = [riscv_instr_group_t.RV32I]
-        self.implemented_csr = []
+        self.shift_t = 2 ** 32 - 1
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:75]: similarly, should this value be a BitArray?
We should still be able to use the binary operators on BitArray structures as well.

File: pygen/pygen_src/riscv_instr_gen_config.py [hunk 4]
```diff
@@ -20,13 +20,11 @@
...
     riscv_instr_group_t
 from pygen_src.target.rv32i import riscv_core_setting as rcs
-logging.basicConfig(format="%(asctime)s %(filename)s %(lineno)s %(levelname)s %(message)s",
-                    level=logging.DEBUG)
 
 
 class riscv_instr_gen_config:
     def __init__(self):
-        self.main_program_instr_cnt = 0  # count of main_prog
+        self.main_program_instr_cnt = 100  # count of main_prog
```
Comment (udinator) [inline pygen/pygen_src/riscv_instr_gen_config.py:27]: please add a TODO here to enable cmdline argument support

File: pygen/pygen_src/riscv_instr_sequence.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        self.program_stack_len = 0  # Stack space allocated for this program
+        self.directed_instr = []    # List of all directed instruction stream
+        self.illegal_instr_pct = 0  # Percentage of illegal instructions
+        self.hint_instr_pct = 0     # Percentage of hint instructions
+
+    def gen_instr(self, is_main_program, no_branch = 1):
+        self.is_main_program = is_main_program
+        self.instr_stream.initialize_instr_list(self.instr_cnt)
+        logging.info("Start generating %d instruction" % len(self.instr_stream.instr_list))
+        self.instr_stream.gen_instr(no_branch = no_branch, no_load_store=1,
```
Comment (udinator) [inline pygen/pygen_src/riscv_instr_sequence.py:37]: nit: please make the spacing around `=` consistent

File: pygen/pygen_src/riscv_instr_pkg.py [hunk 6]
```diff
@@ -1,1084 +1,1207 @@
...
+    MHPMCOUNTER22H = 0xB9
+    MHPMCOUNTER23H = 0xB9
+    MHPMCOUNTER24H = 0xB9
+    MHPMCOUNTER25H = 0xB9
+    MHPMCOUNTER26H = 0xB9
+    MHPMCOUNTER27H = 0xB9
+    MHPMCOUNTER28H = 0xB9
+    MHPMCOUNTER29H = 0xB9
+    MHPMCOUNTER30H = 0xB9
+    MHPMCOUNTER31H = 0xB9
```
Comment (udinator) [inline pygen/pygen_src/riscv_instr_pkg.py:912]: I think these should be `0xB8A` through `0xB9F` ?

File: pygen/pygen_src/riscv_instr_pkg.py [hunk 7]
```diff
@@ -1,1084 +1,1207 @@
...
+    ZBE = auto()
+    ZBF = auto()
+    ZBC = auto()
+    ZBR = auto()
+    ZBM = auto()
+    ZBT = auto()
+    ZB_TMP = auto()
+
+
+class all_gpr(Enum):
```
Comment (udinator) [inline pygen/pygen_src/riscv_instr_pkg.py:1100]: this one can be deleted, we never use this inside the SV generator

File: pygen/pygen_src/riscv_instr_stream.py [hunk 8]
```diff
@@ -1,237 +1,229 @@
...
+            self.instr_list.pop()
+            if len(self.instr_list):
+                break
+
+    def randomize_instr(self, instr, is_in_debug=0, disable_dist=0):
+        exclude_instr = []
+        is_SP_in_reserved_rd = riscv_reg_t.SP in self.reserved_rd
+        is_SP_in_reserved_regs = riscv_reg_t.SP in cfg.reserved_regs
+        is_SP_in_avail_regs = riscv_reg_t.SP in self.avail_regs
+        if ((is_SP_in_reserved_rd or is_SP_in_reserved_regs) or (not is_SP_in_avail_regs)):
```
Comment (udinator) [inline pygen/pygen_src/riscv_instr_stream.py:208]: the SV generator has a fourth check `avail_regs.size() > 0` here also, is this needed?

File: pygen/pygen_src/isa/riscv_instr.py [hunk 9]
```diff
@@ -280,8 +273,68 @@ def post_randomize(self):
...
+                if(self.instr_name.name == "NOP"):
+                    asm_str = "nop"
+                elif(self.instr_name.name == "WFI"):
+                    asm_str = "wfi"
+                elif(self.instr_name.name == "FENCE"):
+                    asm_str = "fence"
+                elif(self.instr_name.name == "FENCE_I"):
+                    asm_str = "fence.i"
+                elif(self.category.name == "LOAD"):
+                    asm_str = '{} {}, {}, ({})'.format(
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:294]: should this be `{} {}, {}({})` ?

File: pygen/pygen_src/isa/riscv_instr.py [hunk 10]
```diff
@@ -280,8 +273,68 @@ def post_randomize(self):
...
+                        asm_str, self.rd.name, self.get_imm(), self.rs1.name)
+                elif(self.category.name == "CSR"):
+                    asm_str = '{} {}, 0x{}, {}'.format(
+                        asm_str, self.rd.name, self.csr, self.get_imm())
+                else:
+                    asm_str = '{} {}, {}, {}'.format(
+                        asm_str, self.rd.name, self.rs1.name, self.get_imm())
+            elif self.format.name == "S_FORMAT":
+                if(self.category.name == "STORE"):
+                    asm_str = '{} {}, {}, {}'.format(
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:304]: similarly, should this be `{} {}, {}({})` ?

File: pygen/pygen_src/isa/riscv_instr.py [hunk 11]
```diff
@@ -280,8 +273,68 @@ def post_randomize(self):
...
+                        asm_str, self.rs2.name, self.get_imm(), self.rs1.name)
+                else:
+                    asm_str = '{} {}, {}, {}'.format(
+                        asm_str, self.rs1.name, self.rs2.name, self.get_imm())
+
+            elif self.format.name == "B_FORMAT":
+                if(self.category.name == "STORE"):
+                    asm_str = '{} {}, {}, ({})'.format(
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:312]: `{} {}, {}({})` ?

File: pygen/pygen_src/riscv_asm_program_gen.py [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        pass
+
+    def gen_dummy_csr_write(self):
+        pass
+
+    def init_gpr(self):
+        reg_val = BitArray(uint = 0, length = pkg_ins.DATA_WIDTH)
+        dist_lst = []
+
+        for dist_val in range(5):
```
Comment (udinator) [inline pygen/pygen_src/riscv_asm_program_gen.py:194]: maybe move the `dist_lst.append(reg_val)` outside of the `if...elif...else` block to avoid duplication?
```python
if dist_val == 0:
  ...
...
if dist_val == 3:
  ...
else:
  ...
dist_lst.append(reg_val)
```

File: pygen/pygen_src/riscv_asm_program_gen.py [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+        subprocess.run(["mkdir", "-p", "out/asm_tests"])
+        file = open("./out/asm_tests/{}".format(test_name), "w+")
+        for items in self.instr_stream:
+            file.write("{}\n".format(items))
+
+        file.close()
+        logging.info("%0s is generated", test_name)
+
+    def gen_signature_handshake(self, instr, signature_type, core_status = "INITIALIZED",
+                                test_result = "TEST_FAIL", csr = "MSCRATCH", addr_lebel = ""):
```
Comment (udinator) [inline pygen/pygen_src/riscv_asm_program_gen.py:392]: nit: `lebel` -> `label`

File: pygen/pygen_src/riscv_instr_pkg.py (hunk 6 above)
Comment (aneels3) [reply pygen/pygen_src/riscv_instr_pkg.py:912]: Yes

File: pygen/pygen_src/riscv_instr_stream.py (hunk 8 above)
Comment (aneels3) [reply pygen/pygen_src/riscv_instr_stream.py]: Yes, We missed that one

File: pygen/pygen_src/isa/riscv_instr.py (hunk 9 above)
Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:294]: Yeah, got it!

File: pygen/pygen_src/isa/riscv_instr.py (hunk 2 above)
Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:56]: We changed self.imm_mask from BitArray to `0xffffffff` because **self.imm** is of **vsc.rand_uint32_t()** type.

We are doing **or** operation in extend_imm function as 
```python3
    self.imm = self.imm_mask | self.imm
```
We were facing some compatibility issue with vsc.rand_uint_t() and BitArray and also on debugging we found that self.imm_mask, the width was greater than 32 bit after shift operation.
That's why we use **self.shift_t** and to maintain the imm_mask width and changed it from BitArray to normal hex value.

Comment (hodjat91) [reply pygen/pygen_src/isa/riscv_instr.py:56]: Thanks @aneels3 , it makes sense. Just a quick suggestion: to be more consistent (and probably get rid of some extra methods), can't we also define self.imm_mask as vsc.uint32_t (just non-random form of vsc.rand_uint32_t) or even vsc.bit_t(width)? Probably in this case we won't face the compatibility issue anymore...

Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:56]: Hi @hodjat91 
Thanks for the suggestion. We tried defining the self.imm_mask as non-random but we are getting the same results. 
I think there is an issue with shift operations in python, it actually padding the zeros after the shift operation. I am not getting why the width is increasing even after defining the width using vsc.uint32_t or vsc.bit_t(32).

Comment (udinator) [reply pygen/pygen_src/isa/riscv_instr.py:56]: This is a good point, I was able to recreate a scenario where the `vsc.uint8_t` could be set to a 9-bit value, etc..., maybe we could open an issue to inquire further about this?

Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:56]: Sure @udinator 
I will open the issue for this.

File: pygen/pygen_src/isa/riscv_instr.py (hunk 1 above)
Comment (hodjat91) [reply pygen/pygen_src/isa/riscv_instr.py:24]: Is this fixed @aneels3 ?

Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:24]: No @hodjat91,
As we run the base_test from the subdirectory of pygen_src. I have used sys module for the relative path.
I am not sure how can I move it all directly into an import statement.
Can I fix this later?

Comment (hodjat91) [reply pygen/pygen_src/isa/riscv_instr.py:24]: Hi @aneels3 ,
Don't overthink it :-) I believe what Udi meant was just to list all of `import A` statements first (on top of the file, after any module comments and docstrings) and then start listing all `from B import C` statements. it is generally a good style to import modules in the following orders: Standard library imports (Python built-in modules), then related third party imports (modules that are installed and do not belong to the current application), and finally local application imports (modules that belong to the current application).

Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py:24]: Thanks, @hodjat91 
Got it now.
I will fix it ASAP.

File: pygen/pygen_src/isa/riscv_instr.py [hunk 14]
```diff
@@ -264,120 +261,187 @@ def set_imm_len(self):
...
+                if(self.instr_name == "NOP"):
+                    asm_str = "nop"
+                elif(self.instr_name == "WFI"):
+                    asm_str = "wfi"
+                elif(self.instr_name == "FENCE"):
+                    asm_str = "fence"
+                elif(self.instr_name == "FENCE_I"):
+                    asm_str = "fence.i"
+                elif(self.category.name == "LOAD"):
+                    asm_str = '{} {}, {},({})'.format(
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:305]: nit: I don't think there should be a comma between `{}` and `({})`, here and for the `STORE` instructions below

File: pygen/pygen_src/isa/riscv_instr.py [hunk 15]
```diff
@@ -264,120 +261,187 @@ def set_imm_len(self):
...
+                        asm_str, self.rd.name, self.get_imm(), self.rs1.name)
+                elif(self.category.name == "CSR"):
+                    asm_str = '{} {}, 0x{}, {}'.format(
+                        asm_str, self.rd.name, self.csr, self.get_imm())
+                else:
+                    asm_str = '{} {}, {}, {}'.format(
+                        asm_str, self.rd.name, self.rs1.name, self.get_imm())
+            elif self.format.name == "S_FORMAT":
+                if(self.category.name == "STORE"):
+                    asm_str = '{} {}, {},({})'.format(
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr.py:315]: nit: no comma here either - `{}({})`

Comment (aneels3) [reply pygen/pygen_src/isa/riscv_instr.py]: Missed that one.
Fixed now!

Comment (udinator) [reply pygen/pygen_src/isa/riscv_instr.py]: thank you!

**Comment by aneels3** [comment]
Hi @udinator 
I am thinking to create a separate PR for command line support.
The left out part of this PR (like constraints) will be up once we get support from PyVSC.
Can we merge this one as some of the files are conflicting while creating the PR from other branch?

**Comment by udinator** [comment]
Hi @aneels3 , feel free to merge this PR, looks like we all approved it.

**Comment by aneels3** [comment]
Thanks, @udinator 
I think I can't merge it because of write access.
